packages/workbox-precaching/src/utils/PrecacheStrategy.ts

```diff
@@ -0,0 +1,138 @@
+
+    if (!response) {
+      // Fall back to the network if we don't have a cached response
+      // (perhaps due to manual cache cleanup).
+      if (handler.params &&
+          handler.params.fallbackToNetwork === false) {
+        // This shouldn't normally happen, but there are edge cases:
+        // https://github.com/GoogleChrome/workbox/issues/1441
+        throw new WorkboxError('missing-precache-entry', {
+          cacheName: this.cacheName,
+          url: request.url,
+        });
+      } else {
+        if (process.env.NODE_ENV !== 'production') {
+          logger.warn(`The precached response for ` +
+              `${getFriendlyURL(request.url)} in ${this.cacheName} was not ` +
+              `found. Falling back to the network instead.`);
+        }
+        response = await handler.fetch(request);
+      }
+    }
+
+    if (process.env.NODE_ENV !== 'production') {
+      // Workbox is going to handle the route.
+      // print the routing details to the console.
+      logger.groupCollapsed(`Precaching is responding to: ` +
+          getFriendlyURL(request.url));
+      logger.log(`Serving the precached url: ` +
+          handler.params.cacheKey);
+
+      logger.groupCollapsed(`View request details here.`);
+      logger.log(request);
+      logger.groupEnd();
+
+      logger.groupCollapsed(`View response details here.`);
+      logger.log(response);
+      logger.groupEnd();
+
+      logger.groupEnd();
+    }
+    return response;
+  }
+
+  async _handleInstall(request: Request, handler: StrategyHandler) {
+    const response = await handler.fetchAndCachePut(request);
+
+    // Any time there's no response, consider it a precaching error.
+    let responseSafeToPrecache = Boolean(response);
+
+    // Also consider it an error if the user didn't pass their own
+    // cacheWillUpdate plugin, and the response is a 400+ (note: this means
+    // that by default opaque responses can be precached).
+    if (response && response.status >= 400 &&
+        !this._usesCustomCacheableResponseLogic()) {
+      responseSafeToPrecache = false;
+    }
+
+    if (!responseSafeToPrecache) {
+      // Throwing here will lead to the `install` handler failing, which
+      // we want to do if *any* of the responses aren't safe to cache.
+      throw new WorkboxError('bad-precaching-response', {
+        url: request.url,
+        status: response.status,
+      });
+    }
+
+    return response;
+  }
+
+  /**
+   * Returns true if any users plugins were added containing their own
+   * `cacheWillUpdate` callback.
+   *
+   * This method indicates whether the default cacheable response logic (i.e.
+   * <400, including opaque responses) should be used. If a custom plugin
+   * with a `cacheWillUpdate` callback is passed, then the strategy should
+   * defer to that plugin's logic.
+   *
+   * @private
+   */
+  _usesCustomCacheableResponseLogic(): boolean {
+    for (const plugin of this.plugins) {
```

Comment:
Optional, but I prefer this logic with `some()`:

```
return this.plugins.some((plugin) => plugin.cacheWillUpdate &&
    plugin.cacheWillUpdate !== copyRedirectedResponses);
```

Reply:
Hmmm, this comment actually got me wondering if it'd make more sense to be checking for the presence of a plugin rather than checking for the presence of a callback. I'll update to do that.

Reply:
Updated. I change the logic slightly to be comparing against a plugin instead of the callback.